compiler/rustc_passes/src/dead.rs

```diff
@@ -832,6 +834,19 @@ fn create_and_seed_worklist(
             effective_vis
                 .is_public_at_level(Level::Reachable)
                 .then_some(id)
+                .filter(|&id| match tcx.def_kind(id) {
+                    // Check impls of trait and impl items of trait later in `check_impl_or_impl_item_live`,
+                    // because `Self` may be non-ADTs and refer to local ADTs.
+                    DefKind::Impl { of_trait } => !of_trait,
+                    DefKind::AssocFn => {
+                        let assoc_item = tcx.associated_item(id);
+                        // Add reachable trait items and impl items not of traits
+                        // into the worklist directly.
+                        matches!(assoc_item.container, AssocItemContainer::Trait)
+                            || assoc_item.trait_item_def_id.is_none()
+                    }
+                    _ => true,
+                })
```

Comment:
How is this change related to supporting non-ADTs?
Isn't this reverting a part of changes from #141407?
The logic is again spread randomly between the various analysis stages, what is the principle behind all this, what is filtered away or analyzed at what stage?

Reply:
The logic here is remove `impl ReachableTrait for ReachableTy<PrivateTy>` from the seed worklist. Because the reachable analysis also considers `impl ReachableTrait for ReachableTy<PrivateTy>` reachable for now.

Reply:
`impl ReachableTrait for ReachableTy<PrivateTy>` and `impl ReachableTrait for [PrivateTy]` may indeed be reachable due to type inference, even it's not intuitive.

I need to check though, whether we mark private items as reachable in other, non-impl cases.
If we do, then private impls can probably be treated as non-reachable as well.